### ISSUE TITLE
Allow rule "selector" property to be a matcher object

### ISSUE THREAD

**Issue by WilcoFiers** [dequelabs/axe-core#1350]
Currently axe-core rules can only take CSS selectors in the selector. This is somewhat limiting, and is the cause of us writing a whole bunch of matchers that often do a very similar thing.

### Proposal

Allow objects as a selector, which will be passed to [`axe.commons.matches()`](https://github.com/dequelabs/axe-core/blob/develop/lib/commons/matches/index.js). This gives us a bunch of new capability, most importantly, letting us deal with fallback rules:

```html
<svg role="graphics-symbol img">...</svg>
```

```json
{
  "id": "svg-img-has-alt",
  "selector": {
	"nodeName": "svg",
    "semanticRole": "img"
  }
}
```

Another common pattern it lets us ditch is this `input:not([type="hidden"]):not([type="image"]):not(...` type selector, which works if you're only interested in one thing, but gets quite unwieldy when it gets more complex.

**Comment by dylanb** [comment]
Why is this better than a matches function?

**Comment by WilcoFiers** [comment]
I'm hoping to drop the quirks we have in axe-core to support IE for the 5.0 release. That means we'll need to support fallback roles. Doing that all through matchers is going to be a problem, so we should figure out how to make this work.

**Comment by straker** [comment]
One thing we'll definitely want to keep note of is the performance implication of using matches functions for the selectors as right now we've [highly optimized the selector lookup for rules](https://github.com/dequelabs/axe-core/pull/3423) through smart caching, but that doesn't work for matches functions
